lib/posix-user/group.c

```diff
@@ -0,0 +1,309 @@
+
+	*rgid = *egid = *sgid = UK_DEFAULT_GID;
+
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, setresgid, gid_t, rgid, gid_t, egid, gid_t, sgid)
+{
+	/* We allow only UK_DEFAULT_GID */
+	if (unlikely(rgid != UK_DEFAULT_GID ||
+		     egid != UK_DEFAULT_GID ||
+		     sgid != UK_DEFAULT_GID))
+		return -EINVAL;
+
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, setfsgid, uid_t, fsgid)
+{
+	return UK_DEFAULT_GID;
+}
+
+/* not a syscall */
+int initgroups(const char *user __unused, gid_t group __unused)
+{
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, getgroups, int, size, gid_t *, list)
+{
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, setgroups, size_t, size, const gid_t *, list)
+{
+	return 0;
+}
+
+static int getgr_r(struct group *ingrp, struct group *grp, char *buf,
+		   size_t buflen, struct group **result)
+{
+	size_t needed;
+	int i, members;
+
+	UK_ASSERT(grp);
+	UK_ASSERT(buf);
+	UK_ASSERT(result);
+
+	if (unlikely(!ingrp)) {
+		*result = NULL;
+		return 0;
+	}
+
+	/* check if provided buffer is large enough */
+	needed = strlen(ingrp->gr_name) + 1 +
+		 strlen(ingrp->gr_passwd) + 1;
+
+	members = 0;
+	while (ingrp->gr_mem[members])
+		needed += strlen(ingrp->gr_mem[members++]) + 1;
+
+	needed += members + 1 * sizeof(char *);
+
+	if (unlikely(buflen < needed)) {
+		*result = NULL;
+		return ERANGE; /* must be positive */
+	}
+
+	/* set name */
+	grp->gr_name = buf;
+	buf = pu_cpystr(ingrp->gr_name, buf);
+
+	/* set passwd */
+	grp->gr_passwd = buf;
+	buf = pu_cpystr(ingrp->gr_passwd, buf);
+
+	/* set gid */
+	grp->gr_gid = ingrp->gr_gid;
+
+	/* set members */
+	grp->gr_mem = (char **) buf;
+	buf += (members + 1) * sizeof(char *);
+	for (i = 0; i < members; i++) {
+		grp->gr_mem[i] = buf;
+		buf = pu_cpystr(ingrp->gr_mem[i], buf);
+	}
+	grp->gr_mem[members] = NULL;
+
+	*result = grp;
+	return 0;
+}
+
+struct group *getgrnam(const char *name)
+{
+	struct group *res;
+
+	if (name && !strcmp(name, g__.gr_name))
+		res = &g__;
+	else {
+		res = NULL;
+		errno = ENOENT;
+	}
+
+	return res;
+}
+
+int getgrnam_r(const char *name, struct group *grp,
+	       char *buf, size_t buflen, struct group **result)
+{
+	return getgr_r(getgrnam(name), grp, buf, buflen, result);
+}
+
+struct group *getgrgid(gid_t gid)
+{
+	struct group *res;
+
+	if (gid == g__.gr_gid)
+		res = &g__;
+	else {
+		res = NULL;
+		errno = ENOENT;
+	}
+
+	return res;
+}
+
+int getgrgid_r(gid_t gid, struct group *grp, char *buf, size_t buflen,
+	       struct group **result)
+{
+	return getgr_r(getgrgid(gid), grp, buf, buflen, result);
+}
+
+int getgrouplist(const char *user, gid_t group, gid_t *groups, int *ngroups)
+{
+	struct passwd *pwd;
+	int ngrps;
+
+	UK_ASSERT(ngroups);
+	UK_ASSERT(groups);
+
+	pwd = getpwnam(user);
+	if (!pwd)
+		return -1;
+
+	ngrps = (group == pwd->pw_gid) ? 1 : 2;
+
+	/* Check if the grouplist is large enough */
+	if (*ngroups < ngrps) {
+		*ngroups = ngrps;
+		return -1;
+	}
+
+	groups[0] = pwd->pw_gid;
+	if (group != pwd->pw_gid)
+		groups[1] = group;
+
```

Comment:
```suggestion
*ngroups = ngrps;
```
Should we update the value stored in `ngroups` to match the actual number of groups found, even if it is greater or equal to `ngrps`?
From the [getgrouplist manual](https://linux.die.net/man/3/getgrouplist):
```text
The ngroups argument is a value-result argument: on return it always contains the number of groups
found for user, including group; this value may be greater than the number of groups stored in groups.
```

Reply:
Yes, makes sense. The part of the man page describing the result value is a bit unclear but I assume you are right. I updated the code. Very good catch 😃